Step3 - 질문 삭제하기 리팩터링 PR요청드립니다.

sangjaeoh opened the pull request:
안녕하세요 😀

JPA 3단계 질문 삭제하기 리팩터링 PR요청드립니다.

질문 드릴게 하나 있는데 `isOwner()` 메소드를 보면  `user.getId().equals(user.getId())`에서 `user.equals(user)`
이런식으로 바꾸고 User 클래스의 equals도 id로만 비교하게 만들었습니다.

엔티티는 테이블과 매핑한다는 관점에서는 `id`만 일치하면 같다라 생각해서 `id`비교로만 `equals()`를 구현했지만,
엔티티 + 도메인으로 봤을 때는 '다른 값도 비교를 해야하지 않을까?' 라고도 생각이 드는데...

혹시 어떻게 생각하는지 알고싶습니다!

감사합니다 :)

sangjaeoh committed:
test: @DisplayName 명확하게 변경

sangjaeoh committed:
docs: 구현 기능 목록 작성

sangjaeoh committed:
feat: Answers 일급 컬렉션 추가

Answers 추가, Question에 List<Answer> 필드를 Answers로 대체 및 그에 따른 사용 코드 수정

sangjaeoh committed:
feat: Question, Answer 글 삭제 메소드 추가

글 삭제 시 글쓴이와 삭제 유저가 동일한지 확인, 다르면 에러 발생
글 삭제 메소드의 반환 값이 List<DeleteHistory>가 될 수 있도록 변경 필요

sangjaeoh committed:
feat: DeleteHistory 정적 팩토리 메소드 추가

Question, Answer를 받아 DeleteHistory 생성

sangjaeoh committed:
feat: Question, Answer 글 삭제 메소드 반환값 List<DeleteHistory>로 변경

sangjaeoh committed:
refactor: QnaService 리팩터링

sangjaeoh committed:
refactor: 불필요한 try/catch 삭제

dhmin5693 commented:
엔티티는 테이블과 매핑한다는 관점에서는 id만 일치하면 같다라 생각해서 id비교로만 equals()를 구현했지만,
엔티티 + 도메인으로 봤을 때는 '다른 값도 비교를 해야하지 않을까?' 라고도 생각이 드는데...

---

맞습니다! 엔티티는 영속화 되어 있다는 조건만 충족하면 id만으로도 비교해도 충분합니다.
대신 영속화 상태임을 확실하게 보장하는 상황에만 사용하는게 좋습니다.

dhmin5693 commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -30,6 +30,14 @@ public DeleteHistory(ContentType contentType, Long contentId, User deletedBy, Lo
         this.createDate = createDate;
     }
 
+    public static DeleteHistory from(Question question) {
+        return new DeleteHistory(ContentType.QUESTION, question.getId(), question.getWriter(), LocalDateTime.now());
+    }
+
+    public static DeleteHistory from(Answer answer) {
+        return new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriter(), LocalDateTime.now());
+    }
```
코드의 의존 방향에 대해서도 고민하면 추후 유지보수에도 큰 도움이 됩니다.

`DeleteHistory`는 삭제 내역을 저장하고 있는 도메인입니다.
지금이야 질문과 답변 삭제 내역만 저장하지만 추후 다른 도메인에도 삭제 내역이 추가될 수 있습니다.
그리고 이런 도메인은 다른 도메인으로부터 생성되기 때문에 스펙이 고정될 확률이 더 높습니다.
예를 들어 삭제 내역에 갑자기 `title` 같은 필드가 추가되진 않을거에요.

또한 `Question`, `Answer` 에선 `DeleteHistory`를 확실히 생성하고 있습니다.
다만 `DeleteHistory`는 `Question`, `Answer` 에서만 생성된다고 보장되지 않습니다.
추후 어디서 생성될지는 미지의 영역에 놓여있어요.

이럴 때 `DeleteHistory`가 `Answer` 나 `Question`에 의존한다면, 그 둘의 변경에 `DeleteHistory`도 영향을 받을 수 있습니다.
영향도를 최소로 하는 방향으로 작성해주세요.

sangjaeoh replied:
`DeleteHistory`가 `Question`, `Answer` 외의 영역에서 생성될 가능성을 생각하지 못했습니다.
만약 `DeleteHistory`의 스펙이 고정되고, 점차 많은 도메인들로 부터 생성될 가능성이 있다고 한다 가정한다면
`of` 메소드로 모두가 범용으로 사용할 수 있도록 변경하겠습니다! 👍

dhmin5693 commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,44 @@
+    protected Answers() {
+    }
+
+    public void addAnswer(Answer answer) {
```
answer에선 add의 대상이 명확하기 때문에 메소드 명은 `add`로만 해도 괜찮지 않을까요?

sangjaeoh replied:
넵!

dhmin5693 commented on src/test/java/qna/domain/AnswersTest.java:
```diff
@@ -0,0 +1,20 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class AnswersTest {
```
`Answers` 단위 테스트가 부족하다고 느껴집니다.
add, remove 등 public method를 충분히 검증해주세요.

sangjaeoh replied:
`size`, `contains` 등 간단한 기능들이고, 
`delete`같은 경우 `AnswerTest`에서의 `delete 테스트` 만으로 충분하다고 생각하여 테스트를 간과하였습니다. 😅

다만 `add`를 테스트 하면서  `Answer`의 연관관계 편의 메소드 `toQuestion`의
```java
if (!question.getAnswers().contains(this)) {
    question.addAnswer(this);
}
```
부분을 
`Answers`의 `add`에서 해야할지 `Answer`해야할지 고민하는 계기가 되었습니다. 🤔

저는 `Answer`와 `Question`의 편의 메소드이니까`Answer`에서 처리하기로 하였습니다.

dhmin5693 replied:
간단한 기능은 맞습니다.
우리는 이 클래스의 기능을 전부 알고 있지만 이 코드를 처음 보는 사람은 클래스가 아니라 인터페이스처럼 받아들여집니다.
인터페이스라는 시각으로 본다면 `Answers` 내부에 데이터 저장소가 `List`인게 중요할까요?

그렇지 않을겁니다. 
내부는 전혀 모르기 때문에 add, remove 등 "사용자가 기대하는 동작"을 잘 되는지만 기대할거에요.

그런 관점에서 잘 짜여진 테스트는 코드의 정상 동작을 보장하면서도 이방인이 쉽게 길을 찾을 수 있는 이정표, 즉 docs 와 같은 역할도 해줄 수 있습니다.

그리고 이 과정 자체가 극단적인 환경을 두고 연습할수록 효과가 좋습니다.
예를 들면 테스트 커버리지 100% 채우기같은 목표를 두고 진행해보시면 실제 업무를 진행할 때에도 테스트를 작성하지 않으면 괜스레 불안하다던가... 뭐 그런게 있더라구요.

100% 는 조금 과해서 따로 피드백으로 남기지는 않지만 이런 방식으로 목표를 세우고 수행해가시는걸 추천드리고 싶습니다.

sangjaeoh replied:
그렇군요... 너무 제 관점에서 생각했던거 같습니다.
`Answers`를 사용하는 사람은 내부 데이터 필드가 `List`인지 모를테니 그에 따라 `add`와 `remove`가 어떻게 동작하는지
설명하는 `docs`과 같은 역할도 하는 `Test`라고 하면 간단한 기능이라도 테스트 하는게 맞겠네요 👍

dhmin5693 commented on src/test/java/qna/service/QnaServiceTest.java:
```diff
@@ -46,7 +43,6 @@ public void setUp() throws Exception {
-        when(answerRepository.findByQuestionIdAndDeletedFalse(question.getId())).thenReturn(Arrays.asList(answer));
 
         assertThat(question.isDeleted()).isFalse();
         qnaService.deleteQuestion(UserTest.JAVAJIGI, question.getId());
```
클래스 필드로 선언해둔 question을 삭제하고 있습니다.
question은 `@BeforeEach`에서 정의된 데이터구요.
각 테스트는 격리된 상황에서 실행되어야 한다는걸 생각하면 공용 리소스를 삭제하는 행위는 다소 위험할 수 있어요.

테스트를 작성할 때 가장 자연스럽게 하는 실수 중 하나가 `@BeforeEach`에 테스트의 조건을 설정하는 것입니다.
`@BeforeEach` 는 테스트가 실행될 수 있도록 환경을 조성해주는 방향으로 작성해주시는게 좋습니다.
테스트의 조건은 given, when, then의 3단계 순서 중 given에서 맞춰주세요.

테스트의 조건인지, 환경인지 헷갈릴 수 있습니다.
when()이나 given()으로 mocking하는데에 사용한다면 조건입니다.

sangjaeoh replied:
`QnaServiceTest`에서 `question`을 삭제 하였다는 부분이 어딘지 잘 모르겠습니다 😅

```java
when(answerRepository.findByQuestionIdAndDeletedFalse(question.getId())).thenReturn(Arrays.asList(answer));
```
위 코드를 삭제한 이유에 대해서 말씀해 주신거라고 생각한다면

`Answer`, `Question`과 가 다대일 양방향 매핑이 되어있고,
`Question`의 `getAnswers`를 호출한 것으로 `List<Answer>`조회가 가능하기 때문에
불필요하게 `answerRepository`로 `List<Answer>`를 직접 조회하는 부분을 삭제하였습니다.
그리고 그에 따라 더이상 사용하지 않게 된 `answerRepository`도 `QnaServiceTest`에서 삭제하였습니다.

dhmin5693 replied:
제가 설명이 부족했던 것 같네요.
말씀해주신 이유 때문에 이 의견을 남겨둔 건 아니었습니다.

`QnaService::deleteQuestion` 은 `Question`을 논리적으로 삭제하는 로직이에요.
이 메소드 내부에선 매개변수 `question`의 deleted 필드를 true로 변경합니다.

이 테스트에 사용한 question은 클래스 필드 레벨로 지정되어 있습니다.
보통 변수를 필드 레벨로 사용하는 이유는 중복 코드로 인해 발생하는 불편을 해소하고자 하기 위함이에요.
하지만 필드가 변경되면 어떻게 되나요?
그 필드를 사용한 모든 메소드가 영향을 받지 않을까요?

이는 테스트의 독립성에 위배되는 행위입니다.
`delete_성공` 테스트가 끝나고 나면 question의 값이 변경될거에요. (deleted = false -> true)
만약 이 question이 다른 테스트에 사용된다면 데이터가 변경된 채로 시작하게 됩니다.

이런 이유로 테스트에서 삭제할 question은 완전히 격리된 환경에서 제공해주는게 좋습니다.
로컬 변수같은 방법으로요.
`@BeforeEach`에서 테스트의 조건을 설정해주지 말라는 건 이런 의미였습니다.

간혹 삭제 로직 중 이미 지워진 리소스는 오류를 내뿜어야 한다는 스펙을 가지고 있는 경우가 있습니다.
만약 question도 그런 스펙을 가졌다면 어떨까요?
question.deleted 가 true로 전환되고, 다른 테스트에서도 삭제를 시도한다면 분명 예외가 발생하겠죠?
그럼 테스트가 실패하게 됩니다. 원인은 당연히 테스트 조건이 오염되어 그렇습니다.

물론 junit 5의 특징 때문에 한 번에 돌려도 실패하지는 않습니다.
최대한 독립적인 테스트 환경을 구성할 수 있는 방법으로 동작하기 때문입니다.
각 테스트 메소드가 실행되기 전에 테스트 클래스(여기선 `QnaServiceTest`)를 매번 새로 생성하고 실행합니다.

즉, 이 클래스의 테스트를 통째로 실행하면 아래처럼 동작합니다.
```
`QnaServiceTest` 생성 -> `@BeforeEach` 실행 -> `delete_성공` 실행
`QnaServiceTest` 생성 -> `@BeforeEach` 실행 ->  `delete_다른_사람이_쓴_글` 실행
`QnaServiceTest` 생성 -> `@BeforeEach` 실행 ->  `delete_성공_질문자_답변자_같음` 실행
```

junit 5가 테스트의 독립성을 최대한 보장해주고는 있다 해도 사용자의 습관이 가장 중요합니다.
왜 테스트는 격리되어야 하는가?
이 이유는 반드시 이해하고 넘어가시는게 좋아요.

그래서 이 의견을 드리게 되었습니다.

sangjaeoh replied:
자세한 설명 감사합니다!
의견 주신것을 보고 오랜 고민 끝에 답변 드립니다. 
<br/>

말씀해주신 것처럼 `question` 가 클래스 필드이기 때문에 테스트 독립성에 위배될 수 있다는 사실을 알고 있습니다.

다만 아래 설명해주신 것 처럼 `@BeforeAll`와 달리 `@BeforeEach`는 각 테스트 메소드가 시작 되기 전에 실행 되고,
매번 `question`을 `new`로 할당해주므로 `question` 클래스 필드를 사용하는 각 테스트에서도 독립성이 보장 된다고 생각합니다.
<br/>

그리고 말씀하신 것처럼 간혹 테스트 조건이 다른 테스트들과 달라 생길 수 있는 문제에 대해서는
그 테스트에서는 전역  변수를 사용하지 않고, 지역변수로 `question`를 생성하여 사용하면 될 거 같다고 생각합니다.
<br/>

또한 전역 변수로 사용하는 `question`이 변경되는 경우가 있을 수 있다고 말씀해주셨습니다.
`question`의 기본 생성자 값이 달라지거나, 그 테스트에서는 다른 생성자를 사용해야 하거나,
필드의 기본값이 다르게 세팅되거나, 삭제되거나, 추가되거나 등이 있다고 생각됩니다.

하지만 이 때는 `@BeforeEach`를 이용해 전역변수를 사용하던 지역변수에서 사용하던
그 테스트에서 사용하는 값이 변경되었다면, 그 테스트는 수정이 필요하다고 생각합니다.
수정 과정에서 전역변수를 쓸지, 지역변수를 만들어서 쓸지는 그 때 결정해도 되고,
반대로 모든 테스트에서 수정 했어야 될 수도 있었던 일이, 단 한번의 수정으로 해결 되는 경우도 있다고 생각합니다.
<br/>

대다수의 테스트에서 중복으로 사용하는 조건을 전역으로 설정할 수 있는
`@BeforeEach`의 장점도 분명히 고려해 봐야 한다고 생각합니다.
<br/>

물론 PR 의견의 주 목적인, `테스트의 독립성을 최대한 보장하는 습관`을 들이기 위해서
`@BeforeEach`사용을 자제하고, 
`@BeforeEach`는 아주 신중히 두 번, 세 번 생각하여 사용하도록 노력하겠습니다. 🫡

감사합니다 :)

dhmin5693 replied:
테스트의 독립성을 잘 지켜주기 위해 TextFixture를 활용하는 방법도 있습니다.
gradle plugin 으로도 있으니 공부해보시면 도움이 될거에요.

sangjaeoh replied:
좋은 정보 알려주셔서 감사합니다 👍
공부해서 사용해보도록 하겠습니다!

dhmin5693 reviewed (changes requested):
안녕하세요 상재님.

리팩토링 전반적으로 정말 잘 해주셨습니다.
살짝 아쉬운 부분에 의견을 조금 남겨두었으니 확인 후 재요청 부탁드립니다.
미션 종료가 눈 앞으로 다가왔으니 조금만 더 힘내주세요. 💪💪💪

sangjaeoh committed:
refactor: 메소드명 수정 및 DeleteHistory 의존성 줄이기

Answers 클래스의 addAnswer 메소드 명을 add로 변경
DeleteHistory 클래스의 form 메소드 삭제 -> of로 변경, 그에 따른 from 사용하던 코드 변경

sangjaeoh committed:
test: AnswersTest 단위테스트 추가

dhmin5693 reviewed (changes requested):
안녕하세요 상재님!

피드백 나머진 잘 반영해주셨는데이해 안된다고 남겨주신 부분에 답변 달아두었습니다.
읽어보시고 다시 한번 고민해보셨으면 합니다.
어려운 점 있으시면 언제든지 DM 보내주세요.

dhmin5693 reviewed (approved):
피드백 잘 반영해주셨네요 :)
JPA 미션은 이걸로 종료하도록 하겠습니다.

지나고 나면 참 쉬워보이지만 숙련도에 따라 굉장히 어려울 수도 있는 미션이 JPA에요.
그럼에도 정말 잘 해주셔서 뿌듯하네요 😁

상재님의 미션 완주를 위해서 언제나 응원하고 있겠습니다.
수고 많으셨습니다. 🙇‍♂️
